hermes-consumers/src/main/java/pl/allegro/tech/hermes/consumers/HermesConsumers.java

```diff
@@ -22,114 +20,101 @@
+                           Trackers trackers,
+                           List<LogRepository> logRepositories,
+                           MessageSenderProviders messageSenderProviders,
+                           MessageSenderFactory messageSenderFactory,
+                           ConsumerNodesRegistry consumerNodesRegistry,
+                           SupervisorController supervisorController,
+                           MaxRateSupervisor maxRateSupervisor,
+                           ConsumerAssignmentCache assignmentCache,
+                           OAuthClient oAuthHttpClient,
+                           HttpClientsWorkloadReporter httpClientsWorkloadReporter,
+                           ConsumersRuntimeMonitor consumersRuntimeMonitor,
+                           ConfigurableApplicationContext applicationContext) {
+        this.springHooksHandler = springHooksHandler;
+        this.consumerHttpServer = consumerHttpServer;
+        this.trackers = trackers;
         this.logRepositories = logRepositories;
-
-        serviceLocator = createDIContainer(binders);
-
-        trackers = serviceLocator.getService(Trackers.class);
-        consumerHttpServer = serviceLocator.getService(ConsumerHttpServer.class);
-        messageSenderFactory = serviceLocator.getService(MessageSenderFactory.class);
-
-        consumerNodesRegistry = serviceLocator.getService(ConsumerNodesRegistry.class);
-        supervisorController = serviceLocator.getService(SupervisorController.class);
-        maxRateSupervisor = serviceLocator.getService(MaxRateSupervisor.class);
-        assignmentCache = serviceLocator.getService(ConsumerAssignmentCache.class);
-        oAuthHttpClient = serviceLocator.getService(OAuthClient.class);
-        httpClientsWorkloadReporter = serviceLocator.getService(HttpClientsWorkloadReporter.class);
-
-        hooksHandler.addShutdownHook((s) -> {
+        this.messageSenderProviders = messageSenderProviders;
+        this.messageSenderFactory = messageSenderFactory;
+        this.consumerNodesRegistry = consumerNodesRegistry;
+        this.supervisorController = supervisorController;
+        this.maxRateSupervisor = maxRateSupervisor;
+        this.assignmentCache = assignmentCache;
+        this.oAuthHttpClient = oAuthHttpClient;
+        this.httpClientsWorkloadReporter = httpClientsWorkloadReporter;
+        this.consumersRuntimeMonitor = consumersRuntimeMonitor;
+        boolean flushLogsShutdownHookEnabled = true;
+        this.applicationContext = applicationContext;
+
+
+        this.springHooksHandler.addShutdownHook((s) -> {
             try {
                 consumerHttpServer.stop();
                 maxRateSupervisor.stop();
                 assignmentCache.stop();
                 oAuthHttpClient.stop();
                 consumerNodesRegistry.stop();
                 supervisorController.shutdown();
-                s.shutdown();
+                applicationContext.registerShutdownHook();
```

Comment:
The only thing that bugs me is that now/in older implementation the hooks were run in order (or at least we've had the ability to do so). When using init and destroy methods it would be troublesome to ensure such an order. But if it is not our concern, and the order of hooks doesn't really matter then it could be done in the proposed way.

Reply:
Yes, you are right that order can be crucial here - I thought about it for a while before I wrote my first comment.
That said, I would prefer to encapsulate dependencies between beans as close to them as possible, rather than having a single class responsible for knowing in what order to start and stop them.

Buy the way, I even found one place where there is such a dependency, but for now I wouldn't focus on it - we will fix it in the follow up PR.